docs/rfcs/20220119_raft_replication_simplification.md

```diff
@@ -0,0 +1,162 @@
+The `recovery_stm` reads a chunk of data and sends it to the follower 
+in a loop (single request at a time).
+
+## Purpose of the simplification
+
+There are multiple reasons that makes the simplification necessary and demanded, 
+the most important are:
+
+- code simplification
+- unification of approaches
+- better handling of requests with mixed consistency level (current they cause latency spikes)
+
+Also some future features that will benefit from simplification are
+
+- in memory acks
+- leader being remote to data 
+
+
+## Proposed solution
+
+Proposed solution assumes that both currently supported consistency levels will 
+be handled in the same way with use of `replicate_batcher` and `replicate_entries` stm. 
+This approach has several advantages over the one currently 
+used for `leader_ack` consistency level. 
+
+Current implementation of `quorum_acks` replication is robust and pretty 
+well tested. No large code changes are required to use the 
+`replicate_batcher` and `replicate_stm` to handle `leader_acks` replication. 
+The way that the `replicate_stm` handles `append_entries` requests dispatching 
+and leader log writes will allow to make them parallel in future. 
+
+Additionally using `replicate_stm` prevents us from creating the 
+append entries request twice since batches coming from `replicate_batcher` 
+are not first handed of to disk and then read again but rather shared to 
+generate leader disk write and follower `append_entries` requests. 
+This approach significantly reduce the `storage::batch_cache` pressure and may 
+also reduce the disk usage (in some cluster with limited memory cache hit 
+ratio is not always 1.0 even if there are no reads other than reads issued by 
+`replicate_stm`).
+
+### Backpressure
+
+In current `quorum_ack` replication implementation the backpressure is 
+propagated from the followers to the `raft::replicate` caller 
+(Kafka Produce request handler effectively). Currently the backpressure 
+propagation is based on the per follower semaphore since RPC layer 
+lacks backpressure propagation mechanism. 
+When `replicate_entries` stm is unable to acquire follower dispatch units 
+it will wait holding `replicate_batcher` memory reservation lock so that 
+it can not accept writes indefinitely. When `replicate_batcher` memory 
+reservation semaphore is exhausted it prevents first stage of `replicate_stages` 
+from finishing not allowing for more requests to be processed. 
+
+Current `leader_ack` mechanism lacks backpressure propagation. 
+Leader can be arbitrary large number of entries ahead of followers. 
+This solution may be a problem since user has an impression that data are 
+replicated while the data are actually stored only on the leader. 
+
+In the proposed solution we are going to handle back pressure in the same way 
+for both consistency level. This way there will be a guarantee that followers 
+can always keep up with the leader.
+
+
+### Memory usage
```

Comment:
I'm interested in the case where:
- initially we have 3 insync replicas
- an acks=1 write passes through replicate_batcher and into replicate_entries_stm
- one of the followers stops responding or returns an error

Do we keep the batches in memory to do some retries (perhaps until the follower is "officially" out of sync due to heartbeat timeout), or do we immediately drop the request to that peer, declare them out of sync and kick them into recovery mode?

Intuitively I'd kind of rather let the peer stay in the "happy path" until we really decide they're out of sync based on a heartbeat timeout, so that something like a TCP disconnect doesn't knock a node into recovery mode, but that could block up the whole write pipeline if we have a finite memory allowance and we're holding on to batches for the unresponsive-but-not-yet-timed-out follower.

Not sure what the right answer is, just something that jumped out to me as an interesting scenario.